src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/PackageSourceMappingUtility.cs

```diff
@@ -12,53 +12,80 @@ namespace NuGet.PackageManagement.UI
+        /// </summary>
+        /// <param name="userAction"></param>
+        /// <param name="addedPackageIds"></param>
+        /// <param name="sourceMappingProvider"></param>
+        /// <param name="existingPackageSourceMappingSourceItems"></param>
+        /// <param name="countCreatedTopLevelSourceMappings">For Top-Level packages: <c>null</c> if not applicable; 0 if none needed to be added; > 0 is the count of new package source mappings added.</param>
+        /// <param name="countCreatedTransitiveSourceMappings">For Transitive packages: <c>null</c> if not applicable; 0 if none needed to be added; > 0 is the count of new package source mappings added.</param>
         internal static void ConfigureNewPackageSourceMapping(
             UserAction? userAction,
             IReadOnlyList<string>? addedPackageIds,
             PackageSourceMappingProvider sourceMappingProvider,
-            IReadOnlyList<PackageSourceMappingSourceItem> existingPackageSourceMappingSourceItems)
+            IReadOnlyList<PackageSourceMappingSourceItem> existingPackageSourceMappingSourceItems,
+            out int? countCreatedTopLevelSourceMappings,
+            out int? countCreatedTransitiveSourceMappings)
         {
+            countCreatedTopLevelSourceMappings = null;
+            countCreatedTransitiveSourceMappings = null;
+
             if (userAction?.SelectedSourceName is null || addedPackageIds is null)
             {
                 return;
             }
 
+            countCreatedTopLevelSourceMappings = 0;
+            countCreatedTransitiveSourceMappings = 0;
+
+            string topLevelPackageId = userAction.PackageId;
             Dictionary<string, IReadOnlyList<string>> patternsReadOnly = existingPackageSourceMappingSourceItems
                 .ToDictionary(pair => pair.Key, pair => (IReadOnlyList<string>)(pair.Patterns.Select(p => p.Pattern).ToList()));
 
             PackageSourceMapping packageSourceMapping = new(patternsReadOnly);
 
             // Expand all patterns/globs so we can later check if this package ID was already mapped.
-            List<string> addedPackageIdsWithoutExistingMappings = new();
+            List<string> addedPackageIdsWithoutExistingMappings = new(capacity: addedPackageIds.Count + 1);
+
             foreach (string addedPackageId in addedPackageIds)
             {
                 IReadOnlyList<string> configuredSource = packageSourceMapping.GetConfiguredPackageSources(addedPackageId);
-                if (configuredSource.Count == 0)
+
+                // Top-level package was looked up.
+                if (addedPackageId == topLevelPackageId)
+                {
+                    // The top-level package is not already mapped to the selected source.
+                    if (configuredSource.Count == 0 || !configuredSource.Contains(userAction.SelectedSourceName))
+                    {
+                        countCreatedTopLevelSourceMappings = 1;
```

Comment:
The vision I have here is when multiple top-levels can be passed in and mapped, the count will become:
```suggestion
                        countCreatedTopLevelSourceMappings++;
```